### PR TITLE
Fix for Issue 113: sendmessage Too Large, Uncatchable Exception

### PR DESCRIPTION
**Pull request by awwad**
Please see Issue #113 for explanation.

Previous to this, trying to sendmessage with a message >9216 bytes on OS X or >65507 on other hosts would crash the sandbox with an uncatchable exception.

This PR raises a RepyArgumentError from sendmessage if the message size is >9216 bytes (for consistency across platforms).

This PR also adds a unit test that addresses these errors. It's a little ugly, but it's r2py. LMK if there's a nicer way to do it. (:

### CODE REVIEW

**Comment by XuefengHuang** [comment]
There is my test result on OpenWrt Chaos Calmer 15.05.1 (WiFi router is TL-WDR3600).
```
root@OpenWrt:/tmp/RUNNABLE# python utf.py -f ut_repyv2api_udp_test_datagram_size
s.r2py
Testing module: repyv2api
	Running: ut_repyv2api_udp_test_datagram_sizes.r2py          [ PASS ]
root@OpenWrt:/tmp/RUNNABLE# python repy.py restrictions.default ut_repyv2api_udp
_test_datagram_sizes.r2py
Message length 1000 TEST OK: sends as expected. 
Message length 1600 TEST OK: sends as expected. 
Message length 9216 TEST OK: sends as expected. 
Message length 9217 TEST OK: causes a RepyArgumentError as expected. 
Message length 10000 TEST OK: causes a RepyArgumentError as expected. 
Message length 65507 TEST OK: causes a RepyArgumentError as expected. 
Message length 65508 TEST OK: causes a RepyArgumentError as expected. 
Message length 65535 TEST OK: causes a RepyArgumentError as expected. 
Message length 65536 TEST OK: causes a RepyArgumentError as expected. 
Message length 100000 TEST OK: causes a RepyArgumentError as expected. 
Terminated
```

File: emulcomm.py [hunk 1]
```diff
@@ -84,6 +84,10 @@
 allowediplist = []
 cachelock = threading.Lock()  # This allows only a single simultaneous cache update
 
+# Trying to send a UDP datagram larger than this size will generate uncatchable
+# exceptions on some systems (OS X). See repy_v2 issue 113.
+MAX_ALLOWABLE_DGRAM_SIZE = 9216
```
Comment (vladimir-v-diaz) [inline emulcomm.py]: I guess this is in bytes?

File: testsV2/ut_repyv2api_udp_test_datagram_sizes.r2py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  except RepyArgumentError:
+    if length in lengths_BAD:
+      log("TEST OK: causes a RepyArgumentError as expected.")
+    else:
+      log("TEST FAIL: should succeed, but instead raises RepyArgumentError:\n")
+      raise
+  except Exception, e:
+    log("TEST FAIL: unexpectedly raises ", type(e), repr(e))
+    raise
+  else:
```
Comment (vladimir-v-diaz) [inline testsV2/ut_repyv2api_udp_test_datagram_sizes.r2py]: For valid messages that don't exceed the UDP datagram limit, I was expecting a check against the value returned by sendmessage().

File: testsV2/ut_repyv2api_udp_test_datagram_sizes.r2py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    "Bad test set: same test case in both OK and BAD set."
+
+ip = getmyip()
+port1, port2 = list(getresources()[0]["messport"])[0:2]
+
+for length in lengths_OK + lengths_BAD:
+  log("Message length", length)
+  try:
+    sendmessage(ip, port1, length * ".", ip, port2)
+  except RepyArgumentError:
```
Comment (vladimir-v-diaz) [inline testsV2/ut_repyv2api_udp_test_datagram_sizes.r2py]: I find it easier to read try-except-else clauses when there are blank lines between clauses.

For example:
```Python
try:
  # code

except:
  # code

else:
  # code

Comment (vladimir-v-diaz) [inline testsV2/ut_repyv2api_udp_test_datagram_sizes.r2py]: Although it is reasonably certain that `ip`, `port1`, and `port2` are valid values here, sendmessage() can raise a `RepyArgumentError` even if the message argument doesn't exceed the UDP datagram limit.

File: testsV2/ut_repyv2api_udp_test_datagram_sizes.r2py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+     -20 bytes for the IPv4 header.
+
+The limit on OS X is instead 9216, which you can see by executing
+     ` sysctl net.inet.udp.maxdgram `
+on an OS X system. Because we value consistency across platforms and think that
+behavior should be indistinguishable across platforms, we apply this limitation
+to all platforms, so 9217 and above should fail via RepyArgumentError from
+sendmessage.
+
+IPv6 jumbograms allow for UDP datagrams larger than that, but we ignore
```
Comment (vladimir-v-diaz) [inline testsV2/ut_repyv2api_udp_test_datagram_sizes.r2py]: This bit of information might be good to mention in `emulcomm.py`, possibly as a TODO or limitation of the current implementation.  Developers / reviewers are likely to read the comments in non-test modules, so it's good to get into the habit of having this sort of information in those modules instead.

**Comment by vladimir-v-diaz** [comment]
The following comment is not strictly an issue with this pull request..

One nitpick I have with the way in which test modules are named and organized is that test conditions for particular methods/functions are scattered throughout.   That is, if I want to find all the tests for sendmessage(), I have to search and consult multiple test modules, and the modules that do relate to sendmessage() have import statements like "pragma repy" and are named `ut_repyv2api_udp_test_datagram_sizes.r2py`  I might just be used to the convention employed by other testing frameworks where one test module usually focuses on one unit.

This is just a comment I wanted to bring up as an "external" reviewer.

File: emulcomm.py (hunk 1 above)
Comment (aaaaalbert) [reply emulcomm.py]: Yes; comment adapted to mention this explicitly.

File: testsV2/ut_repyv2api_udp_test_datagram_sizes.r2py (hunk 2 above)
Comment (aaaaalbert) [reply testsV2/ut_repyv2api_udp_test_datagram_sizes.r2py]: Good suggestion; added.

File: testsV2/ut_repyv2api_udp_test_datagram_sizes.r2py (hunk 3 above)
Comment (aaaaalbert) [reply testsV2/ut_repyv2api_udp_test_datagram_sizes.r2py]: Worth mentioning as an ssllabs style element?

File: testsV2/ut_repyv2api_udp_test_datagram_sizes.r2py (hunk 4 above)
Comment (aaaaalbert) [reply testsV2/ut_repyv2api_udp_test_datagram_sizes.r2py]: Great point; reproduced in `emulcomm` now.

**Pull request merged**